Tutorial 10: Account Updates Tutorial (#313)
o1-labs/docs2: docs/zkapps/tutorials/10-account-updates.mdx

---
title: 'Tutorial 10: Account Updates'
hide_title: true
sidebar_label: 'Tutorial 10: Account Updates'
---

:::info

Please note that zkApp programmability is not yet available on Mina Mainnet, but
zkApps can now be deployed to Berkeley Testnet.

:::

:::note

This tutorial was last tested with [SnarkyJS](https://www.npmjs.com/package/snarkyjs) 0.9.3.

:::

# Tutorial 10: Account Updates

## Permissions, Preconditions, and Composability

> ## Overview:

Reviewer: ```suggestion
## Overview
```

> Each zkApp transaction constructed in SnarkyJS is composed of [AccountUpdates](../snarkyjs-reference/classes/AccountUpdate), a set of instructions for the Mina network.

Reviewer: ```suggestion
Each zkApp transaction constructed by SnarkyJS is composed of one or more [AccountUpdates](../snarkyjs-reference/classes/AccountUpdate), which are a set of instructions for the Mina network to perform, such as altering on-chain state, emitting an event, etc.
```

Each AccountUpdate can make assertions about its account, apply updates to its account, as well as make assertions about its child AccountUpdates.

> Transactions are structured as a list of trees of AccountUpdates - applied with a ["pre-order"](https://en.wikipedia.org/wiki/Tree_traversal) traversal.

Reviewer: ```suggestion
Transactions are structured as a list of trees of AccountUpdates applied with a ["pre-order" traversal](https://en.wikipedia.org/wiki/Tree_traversal).
```

> Many of the core features of zkApps - permissions, preconditions, composability, and tokens - are each implemented through the use of AccountUpdates. In this tutorial, you learn all of these except tokens and assertions on account update layouts.

Reviewer: ```suggestion
Many of the core features of zkApps—permissions, preconditions, composability, and tokens—are implemented using AccountUpdates. In this tutorial, you learn all of these, except tokens and assertions on account update layouts.
```

Reviewer: 'layouts' is a new term that is ambiguous. Could we rephrase or remove that word?

Reviewer: Update: it's mentioned below, but not prior to this point in the tutorial.

> To learn more, see these advanced SnarkyJS docs:

Reviewer: ```suggestion
To learn more, see these SnarkyJS docs:
```

Suggestion to not emphasize `advanced`. It's primarily called that because it goes deeper on topics, but aren't advanced topics, per se.

* [Permissions](../advanced-snarkyjs/permissions)
* [Preconditions](../advanced-snarkyjs/on-chain-values)

## AccountUpdate contents

Each [AccountUpdate](../snarkyjs-reference/classes/AccountUpdate) has these components:

> * `PublicKey`: The account address for the account update

Reviewer: ```suggestion
* `PublicKey`: The address of the zkApp account affected by this account update
```

> * `TokenId`: By default, the Mina TokenId. Together, TokenId and PublicKey uniquely determine the account.

Reviewer: ```suggestion
* `TokenId`: A unique hash representing the custom token. Defaults to the MINA TokenId (`1`). Together, PublicKey and TokenId uniquely identify an account on Mina Protocol.
```

accurate to say 'hash'?

> * `Preconditions`: What needs to be true for the AccountUpdate to be applied. Corresponds to assertions in a SnarkyJS method.

Reviewer: ```suggestion
* `Preconditions`: Conditions that must be true for the AccountUpdate to be applied. Corresponds to assertions in a SnarkyJS method.
```

Does it always or do some create constraints?

> * `Updates`: Things that are changed by the AccountUpdate. Can include the zkApp state, permissions, and verification key.

Reviewer: ```suggestion
* `Updates`: Things changed by the AccountUpdate, such as include the zkApp state, permissions, and verification key.
```

Rephrased because it's not a fully inclusive list.

> * `Balance change`: Any changes to the balance

Reviewer: ```suggestion
* `BalanceChange`: Any changes to the balance
```
The others use PascalCasing. Is this correct in PascalCase as well?

* `Authorization`: How the zkApp is authorized; can be either a proof (corresponding to the verification key on the account) or a signature.

There are some other components we'll discuss in a later tutorial:

* `MayUseToken`: Whether the zkApp has permissions to manipulate its token.
* `Layout`: Allows for assertions about the structure of an AccountUpdate.

## AccountUpdates for a non-user-upgradable zkApp

Now, you can start building an example zkApp to explore permissions, preconditions, and composability with AccountUpdates.

> To visualize transactions, we'll be using the library `mina-transaction-visualizer`. To use this in your own zkApp, install with npm with `npm install mina-transaction-visualizer --save`.

Reviewer: ```suggestion
To visualize transactions, we use the library `mina-transaction-visualizer`. To use this in your own zkApp, install with npm with `npm install mina-transaction-visualizer --save`.
```

> You can find the full source code for this [Account Updates](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/10-account-updates) tutorial in the codebase repo.

Reviewer: ```suggestion
You can find the full [source code for this Account Updates tutorial](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/10-account-updates) in the `examples` directory within the Mina docs repo.
```

    this.account.provedState.get().assertFalse();

    super.init();
    this.num.set(Field(1));
    this.calls.set(Field(0));
  }

  ...
```

This code configures the zkApp as described and initializes the zkApp with the values you want. 

:::caution
Note you assert that `provedState` is `false` in `init()` to ensure that `init()` cannot be called again after the zkApp is set up. Without this, your zkApp could be reset by anyone calling the `init()` method on your zkApp and is recommended for most zkApps. 
:::

Next, add two functions:

```typescript
  ...

  @method add(incrementBy: Field) {
    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertTrue();

    const num = this.num.get();
    this.num.assertEquals(num);
    this.num.set(num.add(incrementBy));

    this.incrementCalls();
  }

  @method incrementCalls() {
    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertTrue();

    const calls = this.calls.get();
    this.calls.assertEquals(calls);
    this.calls.set(calls.add(Field(1)));
  }

  ...
```

:::caution

> In each of these methods, you are also asserting `provedState` is `true` to ensure the zkApp was initialized as expected because `provedState` becomes true after `init` is called and is recommended for most zkApps.

Reviewer: ```suggestion
In each of these methods, you are also asserting `provedState` is `true` to ensure the zkApp was initialized as expected because `provedState` becomes true after `init()` is invoked, and is recommended for most zkApps.
```

:::

> Note you are also having the `add()` method call the `incrementCalls()` method, to see how this is reflected in the `add()` transaction's AccountUpdate structure.

Reviewer: > are also having

Would be good to rephrase this

Finally, add one more function, `callSecondary()` that calls a different zkApp:

```typescript
  ...

  @method callSecondary(secondaryAddr: PublicKey) {
    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertTrue();

    const secondaryContract = new SecondaryZkApp(secondaryAddr);

    const num = this.num.get();
    this.num.assertEquals(num);

    secondaryContract.add(num);

    // NOTE this gets the state at the start of the transaction
    this.num.set(secondaryContract.num.get());

    this.incrementCalls();
  }
}
```

> This takes the address of our other zkApp, `SecondaryZkApp`, and calls a method on it. Note that the impact of calling that method occurs after this set of AccountUpdates - so when you call `secondaryContract.num.get()` that gets the value before this transaction is applied.

Reviewer: ```suggestion
This takes the address of our other zkApp, `SecondaryZkApp`, and calls a method on it. Note that the impact of calling that method occurs after this set of AccountUpdates—so when you call `secondaryContract.num.get()`, it gets the value before this transaction is applied.
```

> Last, to look briefly at [SecondaryZkApp.ts](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/10-account-updates/src/SecondaryZkApp.ts), you have contents:

Reviewer: ```suggestion
Last, to look briefly at [SecondaryZkApp.ts](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/10-account-updates/src/SecondaryZkApp.ts) containing:
```

```typescript
export class SecondaryZkApp extends SmartContract {
  @state(Field) num = State<Field>();

  @method init() {
    super.init();

    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertFalse();

    this.num.set(Field(12));
  }

  @method add(incrementBy: Field) {
    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertTrue();

    const num = this.num.get();
    this.num.assertEquals(num);
    this.num.set(num.add(incrementBy));
  }
}

```

> You declare functions for initializing the account and the `add` method that is called from the earlier `ProofOnlyZkApp`.

Reviewer: ```suggestion
You declare functions for initializing the account and the `add()` method that is called from the earlier `ProofOnlyZkApp`.
```

### Running our Smart Contracts and Visualizing the AccountUpdates

Now it's time to learn about the [main.ts](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/10-account-updates/src/main.ts) file that creates transactions with the earlier smart contracts and the account update visualizations it creates.

First, import the transaction visualizer:

```typescript
...
import { showTxn, saveTxn, printTxn } from 'mina-transaction-visualizer';
...
```

This provides three functions:

```typescript
// creates a png file of a transaction, and opens it in a local image viewer
async showTxn(txn: Mina.Transaction, name: string, legend: Legend)

// creates a png file of a transaction, and saves it to a path
saveTxn(txn: Mina.Transaction, name: string, legend: Legend, path: string)

// prints a nicely formatted view of a transaction

> showTxn(txn: Mina.Transaction, name: string, legend: Legend)

Reviewer 2: ```suggestion
printTxn(txn: Mina.Transaction, name: string, legend: Legend)
```

// with legend type, to replace public keys with human readable strings:
type Legend = { [pk: string]: string };

```

> Now build the legend as follows:

Reviewer: ```suggestion
Now define the legend as follows:
```

```typescript
  const legend = {
    [proofsOnlyAddr.toBase58()]: 'proofsOnlyZkApp',
    [secondaryAddr.toBase58()]: 'secondaryZkApp',
    [deployerAccount.toPublicKey().toBase58()]: 'deployer',
  };
```

Then deploy your smart contracts as follows and visualize the transaction:

```typescript
  const deployTxn = await Mina.transaction(deployerAccount, () => {
    AccountUpdate.fundNewAccount(deployerAccount, 2);
    proofsOnlyInstance.deploy();
    secondaryInstance.deploy();
  });

  await deploy_txn.prove();
  deploy_txn.sign([ deployerKey, proofsOnlySk, secondarySk ]);

  await showTxn(deploy_txn, 'deploy_txn', legend);

  await deploy_txn.send();
```

This yields the following visualization of `deploy_txn`. 

<img src="/img/tutorial-10-deploy_txn.png" width="100%" />

> This is best opened in a new tab to read the contents ([link](/img/tutorial-10-deploy_txn.png)).

Reviewer: ```suggestion
This [visualization](/img/tutorial-10-deploy_txn.png) is best viewed in a new tab.
```

> The deploy transaction includes 5 accountUpdates. In order from left to right;

Reviewer: ```suggestion
The deploy transaction includes 5 accountUpdates represented as ovals. Described from left to right;
```

@es92 should this remain AccountUpdates (PascalCase) for consistency, or is the naming switching to how it's used in JavaScript?

> 1. This accountUpdate takes the new account fee from the deployer for deploying the zkApps. Note the `-2` on the `balanceChange` field

Reviewer: ```suggestion
1. This accountUpdate takes the new account fee from the deployer for deploying the zkApps. Note the `-2` on the `balanceChange` field.
```

Reviewer: I'd suggest referring to it consistently as `accountUpdate` in all 5 bulletpoints, as opposed to switching to 'update' for some.

2. This update deploys the `proofsOnlyZkApp` instance. Note the permissions, all set to the values in the zkApp's deploy field, and the `preconditions` asserting the nonce, so the transaction can't be applied more than once.
3. This update initializes the `proofsOnlyZkApp`. Note the precondition that it can't already be in a proved state.

> 4. This update deploys an instance of `secondaryZkApp`. Note the permissions here are set to default values, in contrast to the deployment in the `proofsOnlyZkApp` example.

Reviewer 3: I don't think we should use accountUpdate @barriebyron, the technical term is "account update" and "update" seems like a reasonable abbreviation of that. We _could_ use `AccountUpdate` here as well

See download link [here](/img/tutorial-10-txn1.png).

Now there are two AccountUpdates, where one is a child of the other. The parent corresponds to the `add` method call - and the child corresponds to the `this.incrementCalls()` call that the parent makes. 

One update is the child because it was called from the parent - which also implies that the parent has the child included as part of its proof. You can read more on parent / child account updates [here](../how-to-write-a-zkapp#payments-and-more-on-public-inputs).

As a reminder, this corresponds to code: 

```typescript
  ...
  @method add(incrementBy: Field) {
    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertTrue();

    const num = this.num.get();
    this.num.assertEquals(num);
    this.num.set(num.add(incrementBy));

    this.incrementCalls();
  }

  @method incrementCalls() {
    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertTrue();

    const calls = this.calls.get();
    this.calls.assertEquals(calls);
    this.calls.set(calls.add(Field(1)));
  }
  ...
```

> You can note in the account updates that in the first update, you're updating the state of `appState[0]` to `5` (matching the `this.num.set` in the `add` contract), and in the second update, you're updating `appState[1]` to `1` (matching the `this.calls.set` in the `incrementCalls` contract).

Reviewer 4: ```suggestion
In the first update, you're updating the state of `appState[0]` to `5` (matching the `this.num.set` in the `add` contract), and in the second update, you're updating `appState[1]` to `1` (matching the `this.calls.set` in the `incrementCalls` contract).
```

> Lastly, call `callSecondary` on your instance of `proofsOnlyZkApp`:

Reviewer 4: ```suggestion
Finally, call `callSecondary` on your instance of `proofsOnlyZkApp`:
```

```typescript
  const txn2 = await Mina.transaction(deployerAccount, () => {
    proofsOnlyInstance.callSecondary(secondaryAddr);
  });

  await txn2.prove();

  await showTxn(txn2, 'txn2', legend);
  await saveTxn(deploy_txn, 'deploy_txn', legend, './txn2.png');

  await txn2.send();
```

> This yields the following visualization of `txn2`

Reviewer: ```suggestion
This returns the following visualization of `txn2`:
```

<img src="/img/tutorial-10-txn2.png" width="100%" />

See download link [here](/img/tutorial-10-txn2.png).

And a quick reminder of the code for `callSecondary`:

```typescript
  @method callSecondary(secondaryAddr: PublicKey) {
    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertTrue();

    const secondaryContract = new SecondaryZkApp(secondaryAddr);

    const num = this.num.get();
    this.num.assertEquals(num);

    secondaryContract.add(num);

>     // NOTE this gets the state at the start of the transaction

Reviewer: @barriebyron maybe we should establish a style for this. Probably not all caps

Reviewer: No guidelines exist for the style of single line code comments. `//` is ideal for the syntax for a single-line JS comment, but not sure about style of the prefix there.

For methods and classes in our code, we use [JS Doc](https://jsdoc.app) comments as our guide, but that won't help here. We haven't defined to greater specificity than that.

    this.num.set(secondaryContract.num.get());

    this.incrementCalls();
  }
```

> This call is producing 3 accountUpdates - one for `callSecondary` (the parent) - one for `secondaryZkApp.add` (the left child) - and one for `incrementCalls` (the right child).

Reviewer 4: ```suggestion
This call produces three accountUpdates:

 - `callSecondary` (the parent)
 - `secondaryZkApp.add` (the left child)
 - `incrementCalls` (the right child)
```

> Following the note in the comment, note that `callSecondary` sets `this.num` to `12` - the value of `secondaryContract` at the "start" of the transaction.

Reviewer: ```suggestion
Where commented in the code, notice that `callSecondary()` sets `this.num` to `12`—the value of `secondaryContract`—at the _start_ of the transaction.
```

> Following the note in the comment, note that `callSecondary` sets `this.num` to `12` - the value of `secondaryContract` at the "start" of the transaction.

Reviewer 4: ```suggestion
As described in the code comment, `callSecondary` sets `this.num` to `12` which is the value of `secondaryContract` at the "start" of the transaction.
```

* [Permissions](../advanced-snarkyjs/permissions)
* [Preconditions](../advanced-snarkyjs/on-chain-values)

## AccountUpdate contents

Each [AccountUpdate](../snarkyjs-reference/classes/AccountUpdate) has several components, the following we'll be discussing here:

* `PublicKey`: The account address for the account update
* `TokenId`: By default the Mina TokenId. Between this and PublicKey, uniquely determines the account
* `Preconditions`: What needs to be true for the AccountUpdate to be applied. Corresponds to assertions in a SnarkyJS method.
* `Updates`: Things that are changed by the AccountUpdate. Can include the zkApp state, permissions, and verification key.
* `Balance change`: Any changes to the balance
* `Authorization`: How the zkApp is authorized; can be either a proof (corresponding to the verification key on the account) or a signature.

There are some other components we'll discuss in a later tutorial:

* `MayUseToken`: Whether or not the zkApp has permissions to manipulate its token.
* `Layout`: Allows for assertions about the structure of an AccountUpdate.

## AccountUpdates for a non-user-upgradable zkApp

> We'll now be building an example zkApp, to explore permissions, preconditions, and composability with AccountUpdates.

Reviewer 4: ```suggestion
Now, you can start building an example zkApp to explore permissions, preconditions, and composability with AccountUpdates.
```

To visualize transactions, we'll be using the library `mina-transaction-visualizer`. To use this in your own zkApp, install with npm with `npm install mina-transaction-visualizer --save`.

> You can find all of the code for this tutorial [here on github](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/10-account-updates).

Reviewer 4: ```suggestion
You can find the full source code for this [Account Updates](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/10-account-updates) tutorial in the codebase repo.
```

### Smart Contracts

> We'll be buliding two smart contracts to use in our example:

Reviewer 4: ```suggestion
You will build two smart contracts to use in your example:
```

First, [ProofsOnlyZkApp.ts](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/10-account-updates/src/ProofsOnlyZkApp.ts).

> This is a zkApp configured to only be modifiable or modifiable via proofs (and for the zkApp we're deploying, is not upgradable once deployed). This means that while someone must own the private key to deploy the zkApp itself - after deployed, only the rules in the zkApp determine how the zkApp functions. The private key is no longer useful for anything.

Reviewer 4: ```suggestion
You will configure this zkApp to be modifiable only by using proofs. For this example, the zkApp is not upgradable after it is deployed. This means that while someone must own the private key to deploy the zkApp itself - after it is deployed, only the rules in the zkApp determine how the zkApp functions. The private key is no longer useful for anything.
```

This zkApp also has functions that call other functions, so we can explore how that impacts a transcation's AccountUpdates.

> We'll start by adding the main contents of the zkApp:

Reviewer 4: ```suggestion
Start by adding the main contents of the zkApp:
```

```typescript

export class ProofsOnlyZkApp extends SmartContract {
  @state(Field) num = State<Field>();
  @state(Field) calls = State<Field>();

  deploy(args: DeployArgs) {
    super.deploy(args);
    this.setPermissions({
      ...Permissions.default(),
      setDelegate: Permissions.proof(),
      setPermissions: Permissions.proof(),
      setVerificationKey: Permissions.proof(),
      setZkappUri: Permissions.proof(),
      setTokenSymbol: Permissions.proof(),
      incrementNonce: Permissions.proof(),
      setVotingFor: Permissions.proof(),
      setTiming: Permissions.proof(),
    });
  }

  @method init() {
    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertFalse();

    super.init();
    this.num.set(Field(1));
    this.calls.set(Field(0));
  }

  ...
```

> This configures the zkApp as described, and initializes the zkApp with the values we'd like. 

Reviewer 4: ```suggestion
This code configures the zkApp as described and initializes the zkApp with the values you want. 
```

:::caution

> Note we assert that `provedState` is `False` in `init`. This ensures that `init` cannot be called again after the zkApp is setup. Without this, our zkApp could be reset by anyone calling the `init` transaction on our zkApp, and is recommended for most zkApps.

Reviewer 4: I edited your suggestion @mitschabaude  and omitted the "in future" because https://developers.google.com/style/future

:::

> Next, we'll add two functions:

Reviewer 4: ```suggestion
Next, add two functions:
```

```typescript
  ...

  @method add(incrementBy: Field) {
    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertTrue();

    const num = this.num.get();
    this.num.assertEquals(num);
    this.num.set(num.add(incrementBy));

    this.incrementCalls();
  }

  @method incrementCalls() {
    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertTrue();

    const calls = this.calls.get();
    this.calls.assertEquals(calls);
    this.calls.set(calls.add(Field(1)));
  }

  ...
```

:::caution

> In each of these, we are also asserting `provedState` is `true`. This ensures the zkApp was initialized as expected, as `provedState` becomes true after `init` is called, and is recommended for most zkApps.

Reviewer 4: ```suggestion
In each of these methods, you are also asserting `provedState` is `true` to ensure the zkApp was initialized as expected because `provedState` becomes true after `init` is called and is recommended for most zkApps.
```

:::

Note we are also having the `add` method call the `incrementCalls` method, to see how this is reflected in an `add` transaction's AccountUpdate structure.

> Lastly, we'll add one more function, `callSecondary`, which calls a different zkApp:

Reviewer 4: ```suggestion
Finally, add one more function, `callSecondary()` that calls a different zkApp:
```

```typescript
  ...

  @method callSecondary(secondaryAddr: PublicKey) {
    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertTrue();

    const secondaryContract = new SecondaryZkApp(secondaryAddr);

    const num = this.num.get();
    this.num.assertEquals(num);

    secondaryContract.add(num);

    // NOTE this gets the state at the start of the transaction
    this.num.set(secondaryContract.num.get());

    this.incrementCalls();
  }
}
```

> This takes the address of our other zkApp, `SecondaryZkApp`, and calls a method on it. Note that the impact of calling that method will occur after this set of AccountUpdates - so when we call `secondaryContract.num.get()`, that will get the value before this transaction was applied.
> 
> Last, to look briefly at [SecondaryZkApp.ts](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/10-account-updates/src/SecondaryZkApp.ts), we have contents:

Reviewer 4: ```suggestion
This takes the address of our other zkApp, `SecondaryZkApp`, and calls a method on it. Note that the impact of calling that method occurs after this set of AccountUpdates - so when you call `secondaryContract.num.get()` that gets the value before this transaction is applied.

Last, to look briefly at [SecondaryZkApp.ts](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/10-account-updates/src/SecondaryZkApp.ts), you have contents:
```

Reviewer 4: @es92 present tense https://developers.google.com/style/tense 

please verify I did not inadvertently change the technical accuracy :)

```typescript
export class SecondaryZkApp extends SmartContract {
  @state(Field) num = State<Field>();

  deploy(args: DeployArgs) {
    super.deploy(args);
    this.setPermissions({
      ...Permissions.default(),
    });
  }

  @method init() {
    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertFalse();

    super.init();
    this.num.set(Field(12));
  }

  @method add(incrementBy: Field) {
    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertTrue();

    const num = this.num.get();
    this.num.assertEquals(num);
    this.num.set(num.add(incrementBy));
  }
}

```

> We declare functions for initializing the account, and a method `add` which is the one that is being called from `ProofOnlyZkApp` above.

Reviewer 4: ```suggestion
You declare functions for initializing the account and the `add` method that is called from the earlier `ProofOnlyZkApp`.
```

### Running our Smart Contracts and Visualizing the AccountUpdates

> We'll now discuss the file [main.ts](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/10-account-updates/src/main.ts), which creates transactions with the above smart contracts, and the account update visualizations it creates.
> 
> First, we'll import the transaction visualizer:

Reviewer 4: ```suggestion
Now it's time to learn about the [main.ts](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/10-account-updates/src/main.ts) file that creates transactions with the earlier smart contracts and the account update visualizations it creates.

First, import the transaction visualizer:
```

```typescript
...
import { showTxn, saveTxn, printTxn } from 'mina-transaction-visualizer';
...
```

> This provides us 3 functions:

Reviewer 4: ```suggestion
This provides three functions:
```

Reviewer 4: tbh, I am not clear that "this" is (the three functions? or the tx?)

```typescript
// creates a png file of a transaction, and opens it in a local image viewer
async showTxn(txn: Mina.Transaction, name: string, legend: Legend)

// creates a png file of a transaction, and saves it to a path
saveTxn(txn: Mina.Transaction, name: string, legend: Legend, path: string)

// prints a nicely formatted view of a transaction
showTxn(txn: Mina.Transaction, name: string, legend: Legend)

// with legend type, to replace public keys with human readable strings:
type Legend = { [pk: string]: string };

```

> We'll build our legend as follows:

Reviewer 4: ```suggestion
Now build the legend as follows:
```

```typescript
  const legend = {
    [proofsOnlyAddr.toBase58()]: 'proofsOnlyZkApp',
    [secondaryAddr.toBase58()]: 'secondaryZkApp',
    [deployerAccount.toPublicKey().toBase58()]: 'deployer',
  };
```

> Then we'll deploy our smart contracts as follows, and visualize the transaction:

Reviewer 4: ```suggestion
Then deploy your smart contracts as follows and visualize the transaction:
```

```typescript
  const deploy_txn = await Mina.transaction(deployerAccount, () => {
    AccountUpdate.fundNewAccount(deployerAccount, 2);
    proofsOnlyInstance.deploy({ zkappKey: proofsOnlySk });
    secondaryInstance.deploy({ zkappKey: secondarySk });
  });

  await deploy_txn.prove();
  deploy_txn.sign([ proofsOnlySk, secondarySk ]);

  await showTxn(deploy_txn, 'deploy_txn', legend);

  await deploy_txn.send();
```

This yields the following visualization of `deploy_txn`. 

> ![](https://raw.githubusercontent.com/es92/zkApp-examples/main/10-account-updates/deploy_txn.png)

Author: Good idea - will aim to do at some point! (Or anyone can submit a PR if they get to it first :) )

> This is best opened in a new tab to read the contents ([link](https://raw.githubusercontent.com/es92/zkApp-examples/main/10-account-updates/deploy_txn.png)).

Reviewer 4: ```suggestion
This visualization is best viewed in a new tab so you can read the contents, like `(https://raw.githubusercontent.com/<username>/zkApp-examples/main/10-account-updates/deploy_txn.png)`
```

Reviewer 4: our tutorial user won't have es92 in their link, right?

The deploy transaction includes 5 accountUpdates. In order from left to right;

1. This accountUpdate takes the new account fee from the deployer for deploying the zkApps. Note the `-2` on the `balanceChange` field
2. This update deploys the `proofsOnlyZkApp` instance. Note the permissions, all set to the values in the zkApp's deploy field, and the `preconditions`, asserting its a new account.
3. This update initializes the `proofsOnlyZkApp`. Note the precondition that it can't already be in a proved state.
4. This update deploys an instance of `secondaryZkApp`. Note the permissions here set to default compared to the deployment in the `proofsOnlyZkApp` example.

> 5. This update initializes the `secondaryZkApp` instance.

Reviewer 3: there's a nuanced thing going on here: because you annotated `init()` with `@method`, there is a separate account update after the deploy one which performs the `init()` logic and has a proof. If you would've left out `@method`, the init stuff would've been part of the deploy update.

Could be interesting to mention that or even make the second method do `init()` without `@method` to highlight the difference. If you think that's worth explaining

> When the transaction is run on-chain, all of these accountUpdates will be checked, and applied. Note that there is no actual "execution" of each AccountUpdate provided - instead each AccountUpdate includes an associated signature or proof, corresponding to the verification key in the zkApp.

Reviewer 3: ```suggestion
When the transaction is run on-chain, all of these account updates will be checked, and applied. Note that there is no actual "execution" of each AccountUpdate provided - instead each AccountUpdate includes an associated signature or proof, corresponding to the verification key in the zkApp.
```

> Next, we'll call `add` on our instance of `proofsOnlyZkApp`:

Reviewer 4: ```suggestion
Next, call `add` on your instance of `proofsOnlyZkApp`:
```

```typescript
  const txn1 = await Mina.transaction(deployerAccount, () => {
    proofsOnlyInstance.add(Field(4));
  });

  await txn1.prove();

  await showTxn(txn1, 'txn1', legend);

  await txn1.send();
```

This yields the following visualization of `txn1`

> ![](https://raw.githubusercontent.com/es92/zkApp-examples/main/10-account-updates/txn1-copy.png)

Reviewer 4: for all generated links, please apply the same feedback on including the username as part of the link

Author: So this was actually for just including a static link to images used in the tutorial - so should be my repo I think (so the images show up).

Probably should eventually be the docs repo, if there is a place images should be going (currently in `/examples/zkapps/10-account-updates`, which not sure reachable for rendering here)

Reviewer 4: the convention to store images is [static/img](https://github.com/o1-labs/docs2/tree/main/static/img) 
<img width="359" alt="image" src="https://user-images.githubusercontent.com/10067215/230656240-7aa8bcbf-733c-4bfa-8ce0-f1ad045113d4.png">

Author: done, thanks

See download link [here](https://raw.githubusercontent.com/es92/zkApp-examples/main/10-account-updates/txn1-copy.png).

> Now we have two AccountUpdates, where one is a child of the other. The parent corresponds to the `add` method call - and the child corresponds to the `this.incrementCalls()` call that the parent makes.

Reviewer 3: I think the tutorial is missing information, or at least a hint, about what it means that "one update is the child of another". I'd recommend to add a sentence about that here. Or maybe you want to save that for the next tutorial?

You could also just link to this section: https://docs.minaprotocol.com/zkapps/how-to-write-a-zkapp#payments-and-more-on-public-inputs
There, the meaning of account update as public inputs is explained in some detail

Reviewer 3: Like, you could say that here, one update is the child of another because it was called by the other one, and that this implies that the parent has the child included as part of its proof.

> Now we have two AccountUpdates, where one is a child of the other. The parent corresponds to the `add` method call - and the child corresponds to the `this.incrementCalls()` call that the parent makes.

Reviewer 4: ```suggestion
Now there are two AccountUpdates, where one is a child of the other. The parent corresponds to the `add` method call - and the child corresponds to the `this.incrementCalls()` call that the parent makes.
```

As a reminder, this corresponds to code: 

```typescript
  ...
  @method add(incrementBy: Field) {
    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertTrue();

    const num = this.num.get();
    this.num.assertEquals(num);
    this.num.set(num.add(incrementBy));

    this.incrementCalls();
  }

  @method incrementCalls() {
    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertTrue();

    const calls = this.calls.get();
    this.calls.assertEquals(calls);
    this.calls.set(calls.add(Field(1)));
  }
  ...
```

> You can note in the accountUpdates that in the first update - we're updating the state of `appState[0]` to `5` (matching the `this.num.set` in the `add` contract) - and in the second update, we're updating `appState[1]` to `1` (matching the `this.calls.set` in the `incrementCalls` contract).

Reviewer 3: ```suggestion
You can note in the account updates that in the first update, we're updating the state of `appState[0]` to `5` (matching the `this.num.set` in the `add` contract), and in the second update, we're updating `appState[1]` to `1` (matching the `this.calls.set` in the `incrementCalls` contract).
```

> Lastly, we'll call `callSecondary` on our instance of `proofsOnlyZkApp`:

Reviewer 4: ```suggestion
Lastly, call `callSecondary` on your instance of `proofsOnlyZkApp`:
```

  await txn2.send();
```

This yields the following visualization of `txn2`

![](https://raw.githubusercontent.com/es92/zkApp-examples/main/10-account-updates/txn2.png)

See download link [here](https://raw.githubusercontent.com/es92/zkApp-examples/main/10-account-updates/txn2.png).

And a quick reminder of the code for `callSecondary`:

```typescript
  @method callSecondary(secondaryAddr: PublicKey) {
    this.account.provedState.assertEquals(this.account.provedState.get());
    this.account.provedState.get().assertTrue();

    const secondaryContract = new SecondaryZkApp(secondaryAddr);

    const num = this.num.get();
    this.num.assertEquals(num);

    secondaryContract.add(num);

    // NOTE this gets the state at the start of the transaction
    this.num.set(secondaryContract.num.get());

    this.incrementCalls();
  }
```

This call is producing 3 accountUpdates - one for `callSecondary` (the parent) - one for `secondaryZkApp.add` (the left child) - and one for `incrementCalls` (the right child).

Following the note in the comment, note that `callSecondary` sets `this.num` to `12` - the value of `secondaryContract` at the "start" of the transaction.

## Conclusion

> We have finished exploring the core features of AccountUpdates, and visualizing the AccountUpdates for a set of transactions. We hope this helps you build more complicated transactions, involving multiple zkApps, as well as understand how SnarkyJS and zkApps is working to enable permissions, preconditions, composability.

Reviewer 4: ```suggestion
Congratulations! You have explored the core features of AccountUpdates and learned about visualizing the AccountUpdates for a set of transactions. You can build more complicated transactions that involve multiple zkApps. This tutorial builds a foundational understanding of how SnarkyJS and zkApps work to enable permissions, preconditions, and composability.
```

> Please see the next tutorial (coming soon) to see how AccountUpdates works with tokens.

Reviewer 4: ```suggestion
```
